Jingle/SCJingleConverter/composition.cpp

```diff
@@ -849,37 +859,137 @@ uint32_t Composition::getMemUsage() {
  *      configured Measure Start and Stop Indices.
  */
 uint32_t Composition::getNumNotes(Channel chan) {
+    if (chan == RIGHT) {
+        return optimizedNotesR.size();
+    } else {
+        return optimizedNotesL.size();
+    }
+}
+
+/**
+ * @brief Composition::optimizeNotes Condenses and trims notes to take up
+ *  less space when downloaded to the Steam Controller.
+ */
+void Composition::optimizeNotes() {
     const uint32_t meas_start_idx = getMeasStartIdx();
     const uint32_t meas_end_idx = getMeasEndIdx();
 
-    uint32_t num_notes = 0;
+    optimizedNotesL.clear();
+    optimizedNotesR.clear();
 
-    Voice* voice = nullptr;
-    switch (chan) {
-    case RIGHT:
-        if (voices.count(voiceStrR)) {
-            voice = &voices[voiceStrR];
+    uint32_t removed_notes = 0;
+
+    // Optimize Notes in Left Channel
+    if (voices.count(voiceStrL)) {
+        // Collect notes into single vector
+        for (uint32_t meas_idx = meas_start_idx; meas_idx <= meas_end_idx; meas_idx++) {
+            Measure& meas = voices[voiceStrL].measures[meas_idx];
+            for (uint32_t notes_idx = 0; notes_idx < meas.notes.size(); notes_idx++) {
+                optimizedNotesL.push_back(meas.notes[notes_idx]);
+            }
         }
-        break;
 
-    case LEFT:
-        if (voices.count(voiceStrL)) {
-            voice = &voices[voiceStrL];
+        // Combine ties
+        for (uint32_t notes_idx = optimizedNotesL.size()-1; notes_idx-- > 0; ) {
+            Note& note = optimizedNotesL[notes_idx];
+
+            if (note.tied && notes_idx < optimizedNotesL.size()-1) {
+                Note& tie = optimizedNotesL[notes_idx+1];
+
+                float note_frequency = note.frequencies.back();
+                float tie_frequency = tie.frequencies.back();
+                if (chordIdxL < note.frequencies.size()) {
+                    note_frequency = note.frequencies[chordIdxL];
+                }
+                if (chordIdxL < tie.frequencies.size()) {
+                    tie_frequency = tie.frequencies[chordIdxL];
+                }
+
+                if (note_frequency == tie_frequency) {
+                    note.length += tie.length;
+                    optimizedNotesL.erase(optimizedNotesL.begin()+notes_idx+1);
+                    removed_notes++;
+                }
+            }
         }
-        break;
-    }
 
-    if (voice) {
-        if (meas_start_idx >= voice->measures.size() || meas_end_idx >= voice->measures.size()) {
-            qDebug() << "Start or End Measure Index out of bounds for Channel in getNumNotes()";
-            return 0;
+        // Remove trailing the trailing rest (should only be one max)
+        if (optimizedNotesL[optimizedNotesL.size()-1].frequencies[chordIdxL] == 0.f) {
+            optimizedNotesL.erase(optimizedNotesL.begin()+optimizedNotesL.size()-1);
         }
+    }
 
+    // Optimize Notes in Right Channel
+    if (voices.count(voiceStrR)) {
```

Comment:
This is similar enough to "// Optimize Notes in Left Channel" block above and code is complex enough that this should be a function (e.g. optimizeNotes(Channel chan)).